examples/runner-custom-terraform-spot-fleet/main.tf

```diff
@@ -0,0 +1,265 @@
+# which is more complex than before
+
+module "gitlab_runner" {
+  source = "../../"
+
+  # CORE OF THIS EXAMPLE
+
+  # Yes, we can pass terraform to gitlab-runner/docker-machine and it apparently works!
+  # If docker_machine_driver is not set to `amazonec2`, you have to set everything manually via
+  # docker_machine_options variable
+  docker_machine_driver = "terraform"
+
+  # Terraform driver has it's own config, check out https://github.com/krzysztof-miemiec/docker-machine-driver-terraform
+  # for more details
+  docker_machine_options = [
+    # Don't forget about trailing slash, which indicates that configuration is a folder and not a single file
+    "terraform-config=/extra-files/runners-config.zip/",
+    "terraform-variables-from=/extra-files/runners-vars.json",
+    "engine-storage-driver=overlay2"
+  ]
+
+  # First, lets include a bunch of files:
+  # - terraform configuration which is used by terraform docker-machine driver to deploy runner instance
+  # - script which is used to download and install terraform and docker-machine-driver-terraform
+  # - configuration file for terraform-created runners
+  extra_files = {
+    "runners-config/main.tf"      = file("${path.module}/worker-config/main.tf")
+    "runners-config/variables.tf" = file("${path.module}/worker-config/variables.tf")
+    "runners-config/outputs.tf"   = file("${path.module}/worker-config/outputs.tf")
+    "docker-machine-terraform.sh" = templatefile("${path.module}/docker-machine-terraform.sh", {
+      terraform_url = "https://releases.hashicorp.com/terraform/0.13.3/terraform_0.13.3_linux_amd64.zip"
+      # This is a fork of @tintoy's docker-machine driver, updated to make TF >=0.12 & <0.14 run with it
+      terraform_driver_url = "https://github.com/krzysztof-miemiec/docker-machine-driver-terraform/releases/download/v0.4.2/linux-amd64.zip"
+    })
+
+    # This corresponds to variables.tf in worker-config folder
+    # Keep in mind that variables set below do not affect instances created using Terraform
+    # and everything has to be set (again) here
+    "runners-vars.json" = jsonencode({
+      image_id                    = data.aws_ami.ubuntu.image_id
+      instance_types              = local.instance_types
+      region                      = var.aws_region
+      subnets                     = local.subnet_ids
+      security_group_ids          = [module.gitlab_runner.runner_sg_id]
+      spot_price                  = "0.2"
+      volume_size                 = 50
+      iam_instance_profile        = aws_iam_instance_profile.runner.name
+      spot_fleet_tagging_role_arn = aws_iam_role.spot_fleet_tagging.arn
+      tags = {
+        "tf-aws-gitlab-runner:instancelifecycle" = "spot:yes"
+      }
+    })
+  }
+
+  # Let's run Terraform & docker-machine-driver-terraform installation script on boot
+  userdata_pre_install = <<SCRIPT
+    chmod +x /extra-files/docker-machine-terraform.sh
```

Comment:
This example suffers a bit. If we change the `docker-machine-terraform.sh` in the `extra_files`, we will indeed get the new version of the files on the machine (via SSM command). But the new version will not be installed.

We want to taint the machine somehow if the `docker-machine-terraform.sh` content changes  so that it's re-made with the new content.

A simple example could be to include a hash of the realised `docker-machine-terraform.sh` file in the `userdata`, perhaps as a comment or something. This will taint the ASG and force the machine replacement.